### PR TITLE
Wizards/UI: refactor SAM deploy, add bunch of tests

### PR DESCRIPTION
**Pull request by JadenSimon**
Apologies for the massive PR. As I refactored the SAM deploy wizard, I noticed a lot of things that I wanted to fix/change with the prompter code. And those changes also came with changes to testing code, as well as the addition of several new tests.

## Major changes
* Wizard prompt bindings can now declare dependencies on other properties. Prompt ordering is resolved based on the constructed dependency graph. Types also reflect the dependencies: if a prompt depends on `foo`, then the prompt state/context will always show `foo` as defined. This added `toposort` as a dependency, though it's a tiny one.
* Removed `CachedPrompter` entirely. Its main purpose was to route the cache provided by the wizard to the appropriate UI element, however, having to extend a class for this feature is pretty heavy. Instead cached item loading can be provided by using partially applied functions with deferred cache binding.`node-object-hash` was added as a dependency, but it's possible to do the hashing ourselves. See `partialCache` + `createCachedFunction` to see how it works.
* Added the idea of an item being 'recently used'. Previously there was only a single mechanism for setting the item which was primarily used by the wizard flow for forward/backwards movement through prompts. This led to issues when trying to set an item as 'recently used' outside the context of a wizard. The new field automatically takes care of prepending the 'recently used' suffix in a _declarative_ way. That is, callers only need to set this field to true per-item and the rest will be handled by downstream logic. 
* Started the process of removing the `Prompter` base class and turn it into an interface. The base class has some nice functionality, though strictly speaking pretty much all of the base logic isn't used by the `Wizard` code and could be turned into standalone ideas. 
* Added basic region caching to the S3 client. It looks like the SDK cache is being cleared because they do not handle partitions, so I added caching logic that correctly handles partitions. The caching also respects changes in 'createTime' which means the bucket may have been deleted and re-created in another region.
* Dropped the `prompter` suffix from common UI implementation files.
* `noItemsFoundItem` can just be a string now.
* `errorItem` can be a string or a callback that resolves an error to a string/item.
* Reworked the S3 bucket prompt to not have buttons to create or select a bucket not in the list. It instead uses the 'filter box' as a way to input things. We automatically check if the bucket exists, and if not the user will be prompted to create the bucket with the given name (the GIF is outdated but it gets the point across):
#### TODO: jump to next prompt after successful create and change the text
Kapture 2021-11-08 at 18 39 40

## Test code updates
* [WizardTester] Added the ability to run a UI tester directly against a wizard prompt via `runPrompt`. Previously wizard flow logic and 'prompt' logic had to be tested separated, now they can be tested together. Currently only quick picks can be tested.
* [WizardTester] Added debug information print-out on test failure. Wizards can have a lot going on, so it's nice to see what the current state was when the test case failed. Example output:
Screen Shot 2021-11-08 at 6 13 29 PM
blue is 'assigned state', green is ready-to-prompt in that order, red is won't prompt
* [QuickPickTester] Minor updates for better debug output. Current plans are to expand this test code to `InputBox` as well as allow for more ways to test things. 

## TODO:
* Finish writing the remainder of tests. There's a few I'm missing still, mostly because the expanded testing functionality allows for more to be tested.
* Have the S3 bucket prompter jump to next prompt after successfully creating a bucket. This logic can easily be reused as well (ECR, tags, etc.)

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by JadenSimon** [comment]
1.44 doesn't appreciate the test code that uses the filter box. It's _probably_ an event not being fired (`onDidChangeValue`). I'll just dig through VS code's source at some point.

File: src/shared/wizards/util.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    furthestStep: string
+}
+
+interface TraceReporter {
+    stop: (result: any) => void
+}
+
+export type WizardTraceResult<TState> = TraceCompleted<TState> | TraceFailed | TraceCancelled
+
+export class WizardTrace<TState> {
```
Comment (JadenSimon) [inline src/shared/wizards/util.ts:48]: This will be used to inject metadata into telemetry. It can capture a lot of information:
* Total duration
* Time spent per-prompt
* What step did the user cancel on?
* Was there an uncaught error? If so, which step?
* What was the furthest step the user got to?
* How many total prompts did the user see?

I doubt everything will be used, but it's a start.

File: src/test/shared/ui/pickerPrompter.test.ts [hunk 2]
```diff
@@ -277,124 +315,104 @@ describe('QuickPickPrompter', function () {
...
+        const input = 'not a number'
+        const validator = async (val: string) => {
+            await new Promise(r => setImmediate(r))
+            return filterBoxInput.validator(val)
+        }
+        const filterBox = { ...filterBoxInput, validator }
+
+        const newPicker = vscode.window.createQuickPick() as DataQuickPick<number>
+        testPrompter = new FilterBoxQuickPickPrompter(newPicker, { filterBoxInput: filterBox })
+        tester = createQuickPickTester(testPrompter, { forceEmits: true })
```
Comment (JadenSimon) [inline src/test/shared/ui/pickerPrompter.test.ts:405]: Setting the `forceEmits` option will trigger the event emitter for certain actions rather than letting VS Code fire the event. The reason for this is that VS Code will trigger the event 'at some point', though this leads to flaky tests. This flag could potentially lead to a deviation in test behavior vs. actual which is why it is false by default.

File: src/shared/clients/s3Client.ts [hunk 3]
```diff
@@ -119,13 +119,33 @@ export interface DeleteBucketRequest {
...
         private readonly regionCode: string,
         private readonly s3Provider: (regionCode: string) => Promise<S3> = createSdkClient,
         private readonly fileStreams: FileStreams = new DefaultFileStreams()
-    ) {}
+    ) {
+        DefaultS3Client.regionCache[partitionId] ??= {}
+    }
+
+    /** Cache bucket region based off partition */
+    private static regionCache: RegionCache = {}
```
Comment (JadenSimon) [inline src/shared/clients/s3Client.ts:142]: Loading S3 buckets was _painfully_ slow with 80+ buckets. It'd take 20-30 seconds to refresh because it would check the region for every bucket every single time. Not good UX. Over-aggressive caching isn't good UX either so we need to base it off partitions + the bucket create time.

**Comment by JadenSimon** [comment]
> codecov/project Successful in 1s — 66.38% (+1.32%) compared to 098744d 

Guess I didn't write enough tests. Gotta get at least 2% 🥇

File: .vscode/launch.json [hunk 4]
```diff
@@ -64,7 +64,8 @@
                 "TEST_REPORT_DIR": "${workspaceFolder}/.test-reports",
                 "TEST_FILE": "null",
                 "AWS_TOOLKIT_IGNORE_WEBPACK_BUNDLE": "true",
-                "NO_COVERAGE": "true"
+                "NO_COVERAGE": "true",
+                "FORCE_COLOR": "2"
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply .vscode/launch.json:68]: Note: one reason for not displaying colors is that systems that can't display colors look worse, e.g. the Github bot output: 
image

[the remark this answers is not in the document]
Comment (JadenSimon) [reply .vscode/launch.json:68]: Good callout. Perhaps only set it for individual tests since that's when it would be most useful anyway.

File: src/shared/wizards/wizardForm.ts [hunk 5]
```diff
@@ -3,87 +3,159 @@
...
+ *       qaz?: string
+ *    }
+ *    baz?: string
+ * }
+ * ```
+ * `M` is used to describe when there is a matched key. Matched keys should return their type,
+ * while non-matches are considered to be unassigned, and thus are potentially undefined.
+ */
+type FindPartialTuple<T, K extends string[], M extends boolean = false> = K['length'] extends 0
+    ? M extends true
```
Comment (bryceitoc9) [inline src/shared/wizards/wizardForm.ts:45]: Why do all of these need to extend types, are they not those types in general?

File: src/shared/wizards/wizardForm.ts [hunk 6]
```diff
@@ -3,87 +3,159 @@
...
+
+export type PrompterProvider<TState, TProp, TDep> = (
+    state: StateWithCache<BoundState<TState, TDep>, TProp>
+) => Prompter<TProp>
+export type BoundState<TState, Bindings> = FindPartialTuple<TState, ExtractPaths<Bindings>>
+
+type DefaultFunction<TState, TProp, TDep> = (state: BoundState<TState, TDep>) => TProp | undefined
+
+/**
+ * Changes all properties enumerated by `K` to be required while leaving everything else
```
Comment (bryceitoc9) [inline src/shared/wizards/wizardForm.ts:20]: I'm going off a LOT of faith that this works, I'm assuming there's no way to test this? Seems like a lot of black magic, and unless this is bulletproof I think we'd have a hell of a time debugging it at a later date. Is this absolutely necessary?

File: src/shared/wizards/wizardForm.ts [hunk 7]
```diff
@@ -3,87 +3,159 @@
...
-import { ExpandWithObject } from '../utilities/tsUtils'
+import * as toposort from 'toposort'
+import { Prompter } from '../ui/prompter'
+import { StateWithCache } from './wizard'
+import { ExpandWithObject, ReduceTuple } from '../utilities/tsUtils'
+
+export type PrompterProvider<TState, TProp, TDep> = (
+    state: StateWithCache<BoundState<TState, TDep>, TProp>
+) => Prompter<TProp>
+export type BoundState<TState, Bindings> = FindPartialTuple<TState, ExtractPaths<Bindings>>
```
Comment (bryceitoc9) [inline src/shared/wizards/wizardForm.ts:15]: Lots of explanation about the `FindPartialTuple` but there's no actual description for what this does with that information and how to use this type, despite the fact that this is the only part that future users will see.

File: src/shared/wizards/wizardForm.ts [hunk 8]
```diff
@@ -3,87 +3,159 @@
...
-export type PrompterProvider<TState, TProp> = (state: StateWithCache<WizardState<TState>, TProp>) => Prompter<TProp>
+// This is basically just a 'type-helper' to retain information about where to find a property
+// relative to a 'root' element
+interface Pathable {
+    path: string[]
+}
 
-type DefaultFunction<TState, TProp> = (state: WizardState<TState>) => TProp | undefined
+// Extracts the 'paths' of a type relative to the root type
+type ExtractPaths<B> = B extends { path: infer K }[] ? (K extends string[] ? K : never) : never
```
Comment (bryceitoc9) [inline src/shared/wizards/wizardForm.ts:66]: What is the expected output of this? At the end of all of this, it looks like you're just returning a series of arrays, is there an easier way to do this?

File: src/shared/wizards/wizardForm.ts [hunk 9]
```diff
@@ -97,25 +169,25 @@ export class WizardForm<TState extends Partial<Record<keyof TState, unknown>>> {
         this.body = this.createWizardForm()
     }
 
+    // note: this solves the dependency tree, so even though it's an accessor it's a pretty expensive one
```
Comment (bryceitoc9) [inline src/shared/wizards/wizardForm.ts:172]: Can this be cached? Can this also have a proper jsdoc as a public getter?

File: src/shared/wizards/wizardForm.ts [hunk 10]
```diff
@@ -97,25 +169,25 @@ export class WizardForm<TState extends Partial<Record<keyof TState, unknown>>> {
...
     public get properties(): string[] {
-        return [...this.formData.keys()].sort(this.compareOrder.bind(this))
+        return this.resolveDependencies()
     }
 
-    public getPrompterProvider(prop: string): PrompterProvider<TState, any> | undefined {
+    public getPrompterProvider(prop: string): PrompterProvider<TState, any, any> | undefined {
```
Comment (bryceitoc9) [inline src/shared/wizards/wizardForm.ts:177]: I feel like this is an awful lot of `any`s for a public function. Can we tie this to the class' types?

File: src/shared/wizards/wizardForm.ts [hunk 11]
```diff
@@ -3,87 +3,159 @@
  * SPDX-License-Identifier: Apache-2.0
```
Comment (bryceitoc9) [inline src/shared/wizards/wizardForm.ts:3]: As a whole, within this file, there is a LOT of logic running through types, which are as far as I know untestable. Is it worth investing this heavily in types? The syntax is obscure, and troubleshooting/editing/general maintenance is not going to be easy, especially in the case of people rotating in/out of the team with less Typescript knowledge.

As a thought experiment: are there other ways to implement this that relies less on the type system and more on testable helper functions? I might be missing something, but it doesn't seem like there's any way to test the output of the toposort, which seems to play heavily into the types here (I very well might be wrong)?

File: src/shared/wizards/util.ts (hunk 1 above)
Comment (JadenSimon) [reply src/shared/wizards/util.ts:48]: Yeah I was going to pull it out, though I wanted to highlight it to show some ideas I had in mind. The changes to the wizard are very minor, but I did need to implement this on my 'main' branch to get an idea of how it'd work.

The step number _does_ mean something, though you're correct that a step identifier is needed. In this case the default identifier (currently implemented) is just the bound property which is easily human readable but is more tightly coupled to the code. Currently just thinking of having a required 'id' field somewhere.

File: src/shared/ui/pickerPrompter.ts [hunk 12]
```diff
@@ -567,62 +601,89 @@ export class QuickPickPrompter<T> extends Prompter<T> {
...
+                                if (validatedValue !== picker.value) {
+                                    // stale validation
+                                    update(picker.value)
+                                    return
+                                }
+                                const inputItem = createItem(result, !!result)
+                                picker.items = [inputItem, ...items]
+                            })
+                        )
+                    }, DEBOUNCE_TIME)
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/ui/pickerPrompter.ts:678]: The timer is started as soon as the user starts typing. Debouncing is achieved by clearing the previous timer. So the code block is not executed as long as the user keeps typing within the debounce interval. There's an inner `if` statement here that needs to check if the scoped-in value is still valid as well.

File: src/shared/utilities/collectionUtils.ts [hunk 13]
```diff
@@ -274,9 +275,154 @@ export function stripUndefined(obj: any): void {
     })
 }
 
-export function isAsyncIterable(obj: any): obj is AsyncIterable<unknown> {
-    return (
-        Object.getOwnPropertySymbols(obj).includes(Symbol.asyncIterator) &&
-        typeof obj[Symbol.asyncIterator] === 'function'
-    )
+export function entries<T extends Record<PropertyKey, any>>(obj: T): [keyof T, T[keyof T]][] {
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/utilities/collectionUtils.ts:278]: Remnant from other branches. I at least was using this at one point, but not anymore. All it does is make `Object.entries` typed correctly. Can be annoying to have to case the keys or values.

File: src/apprunner/wizards/codeRepositoryWizard.ts [hunk 14]
```diff
@@ -150,71 +138,42 @@ function createPortPrompter(): InputBoxPrompter {
...
+    }
 
-        refreshButton.onClick = refresh
+    const prompter = createQuickPick<ConnectionSummary>([], {
+        title: localize('AWS.apprunner.createService.selectConnection.title', 'Select a connection'),
+        buttons: createCommonButtons(apprunnerConnectionHelpUrl),
+        itemLoader: partialCached(itemLoader, client.regionCode),
+        noItemsFoundItem: noConnection,
+        compare: (a, b) => (a.detail ? 1 : b.detail ? -1 : 0),
+        errorItem: localize('AWS.apprunner.createService.selectConnection.failed', 'Failed to list GitHub connections'),
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/apprunner/wizards/codeRepositoryWizard.ts:173]: Only GitHub at the moment.

File: src/shared/clients/s3Client.ts (hunk 3 above)
Comment (JadenSimon) [reply src/shared/clients/s3Client.ts:142]: I wrote the code to handle this exact scenario (and tested it). It will evict the cache and load the correct region since the `createDate` has changed. I'll see if I can make a unit test as well (I think I wrote one but cannot remember).

File: src/shared/ui/buttons.ts [hunk 15]
```diff
@@ -7,17 +7,18 @@ import * as vscode from 'vscode'
...
 const HELP_TOOLTIP = localize('AWS.command.help', 'View Toolkit Documentation')
 
-type WizardButton<T> = QuickInputButton<T | WizardControl> | QuickInputButton<void>
-export type PrompterButtons<T> = readonly WizardButton<T>[]
-
+export type PrompterButtons<T = never, U extends Prompter<T> = any> = readonly QuickInputButton<T, U>[]
 /** Light wrapper around VS Code's buttons, adding a `onClick` callback. */
-export interface QuickInputButton<T> extends vscode.QuickInputButton {
-    onClick?: () => T
+export interface QuickInputButton<T = never, U extends Prompter<T> = any> extends vscode.QuickInputButton {
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/ui/buttons.ts:20]: If you don't specify a type then it should never return anything.

File: src/shared/ui/buttons.ts [hunk 16]
```diff
@@ -7,17 +7,18 @@ import * as vscode from 'vscode'
...
+    onClick?: (prompter: U) => UnionPromise<PromptResult<T> | void>
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/ui/buttons.ts:21]: Button handlers in this case are the prompters. A button will 'inherit' the correct type if using `addButton`
If you specify the type yourself that works too, but you will not know what kind of 'prompter' will be passed to you (just manually type it).

File: src/shared/ui/common/basicExit.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+import { Prompter, PrompterConfiguration, PromptResult } from '../prompter'
+
+class ExitPrompter extends Prompter<boolean> {
+    private prompter: QuickPickPrompter<boolean>
+
+    public get recentItem(): any {
+        return undefined
+    }
+
+    public set recentItem(response: any) {}
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/ui/common/basicExit.ts:17]: Going backwards/forwards through wizards mainly.
Probably should split up the interfaces to hide these implementation details so they're not exposed for the single-use prompt case.

File: src/shared/ui/common/basicExit.ts [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+                { label: localize('AWS.generic.response.no', 'No'), data: false },
+                { label: localize('AWS.generic.response.yes', 'Yes'), data: true },
+            ],
+            {
+                title: localize('AWS.wizards.exit.title', 'Exit wizard?'),
+            }
+        )
+    }
+
+    public override dispose() {
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/ui/common/basicExit.ts:33]: Decorative. Let readers know this is an override.

File: src/shared/ui/common/basicExit.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        this.prompter.dispose()
+        super.dispose()
+    }
+
+    protected async promptUser(config: PrompterConfiguration): Promise<PromptResult<boolean>> {
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/ui/common/basicExit.ts:38]: Yeah I'll add a comment. That's the behavior.

File: src/shared/ui/common/basicExit.ts [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        return await this.prompter.promptControl()
+    }
+}
+
+declare interface ExitPrompterConstructor {
+    (): ExitPrompter
+    new (): ExitPrompter
+}
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/ui/common/basicExit.ts:51]: It's sugaring but not sure if I like it. Messing around with classes that can be called as functions.

File: src/lambda/wizards/samDeployWizard.ts [hunk 21]
```diff
@@ -13,950 +13,282 @@ import * as vscode from 'vscode'
...
-        this.response.region = await this.context.promptUserForRegion(step, this.response.region)
-
-        return this.response.region ? wizardContinue(this.S3_BUCKET) : WIZARD_GOBACK
-    }
+        this.form.region.bindPrompter(() =>
+            createRegionPrompter({
+                title: localize('AWS.samcli.deploy.region.prompt', 'Which AWS Region would you like to deploy to?'),
+                defaultRegion: context.awsContext.getCredentialDefaultRegion(),
+                helpUri: samDeployDocUrl,
+            }).transform(r => r.id)
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/lambda/wizards/samDeployWizard.ts:180]: A lot of downstream logic just uses the region id, though arguably it'd make more sense to pass around `Region` until the last moment.

File: src/lambda/wizards/samDeployWizard.ts [hunk 22]
```diff
@@ -13,950 +13,282 @@ import * as vscode from 'vscode'
...
 
+type CFNTemplate = CloudFormation.Template & {
+    uri: vscode.Uri
+    parameterOverrides?: Map<string, string>
+    missingParameters?: Set<string>
+}
+export const CONFIGURE_PARAMETERS = new Map<string, string>()
+
 export interface SamDeployWizardResponse {
+    missingParameters?: Set<string>
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/lambda/wizards/samDeployWizard.ts:51]: Yeah it can now. It was pretty tough to extract out the old missing params logic so I'm sure there's still some duplication present.

File: src/apprunner/wizards/codeRepositoryWizard.ts [hunk 23]
```diff
@@ -51,38 +45,32 @@ function createRepoPrompter(git: GitExtension): QuickPickPrompter<Remote> {
...
+
+        return branches
+            .filter(b => !!b.name)
+            .map(b => b.name!.split('/').slice(1).join('/'))
+            .map(b => ({ label: b, data: b }))
+    }
+
     const userInputString = localize('AWS.apprunner.createService.customRepo', 'Enter branch name')
-    return createLabelQuickPick(branchItems, {
+    return createLabelQuickPick([], {
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/apprunner/wizards/codeRepositoryWizard.ts:67]: Reworked it to have the loader be the first arg with base items as an option (this was the original goal)

File: src/apprunner/wizards/codeRepositoryWizard.ts [hunk 24]
```diff
@@ -150,71 +138,42 @@ function createPortPrompter(): InputBoxPrompter {
...
+        onClick: vscode.env.openExternal.bind(vscode.env, vscode.Uri.parse(apprunnerConnectionHelpUrl)),
+    }
 
-        refreshButton.onClick = refresh
+    const prompter = createQuickPick<ConnectionSummary>([], {
+        title: localize('AWS.apprunner.createService.selectConnection.title', 'Select a connection'),
+        buttons: createCommonButtons(apprunnerConnectionHelpUrl),
+        itemLoader: partialCached(itemLoader, client.regionCode),
+        noItemsFoundItem: noConnection,
+        compare: (a, b) => (a.detail ? 1 : b.detail ? -1 : 0),
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/apprunner/wizards/codeRepositoryWizard.ts:172]: It's because 'Pending' is placed in the detail. We want pending at the bottom.

File: src/shared/utilities/collectionUtils.ts [hunk 25]
```diff
@@ -3,7 +3,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import './asyncIteratorShim'
+import * as hasher from 'node-object-hash'
+import { SorterOptions } from 'node-object-hash/dist/objectSorter'
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/utilities/collectionUtils.ts:7]: Sometimes modules will not export all of the types/interfaces they use, so we need to import them from their source

File: src/shared/wizards/wizardForm.ts (hunk 11 above)
Comment (JadenSimon) [reply src/shared/wizards/wizardForm.ts:3]: Well this file is like 90% documentation now. The types here are really not that crazy, it just seems weird if you have never dealt with the syntax at all. Will be sitting on this PR until after the next release though. I'll see what I can part out in the meantime to make it easier to digest.

Comment (hunterwerlla) [reply src/shared/wizards/wizardForm.ts:3]: It has been a while since I have contributed to this project, but I want to make a quick comment on this because it seems like a fundamental change in the style of the code base. 

Originally, the toolkit for VSCode used the Typescript type system much more fully than it does today. However, after the people working on it changed significantly, a lot of the uses of the more advanced type system features were removed because they made the code hard to understand. Personally, I feel like I have a good handle on generics and some handle on template metaprogramming in C++, but I find a lot of this code is hard to parse: I'd rather try to achieve the same level of type safety with duplication and more verbosity.

If you believe this code will not require significant changes or debugging in the future, dense code like this is certainly less of a problem. However, in my experience, there is not that much code that does not change in an actively developing and expanding project like this, especially as VSCode itself changes.

File: src/shared/clients/s3Client.ts (hunk 3 above)
Comment (kiiadi) [reply src/shared/clients/s3Client.ts:142]: We should probably do this in all our toolkits

File: src/shared/ui/common/basicExit.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (kiiadi) [reply src/shared/ui/common/basicExit.ts:33]: it also protects against the parent interface changing in future and this implementation not realizing 

probably want to this consistently (maybe as a separate PR) and the linting rule for it : https://www.typescriptlang.org/tsconfig#noImplicitOverride

File: src/shared/wizards/wizardForm.ts [hunk 26]
```diff
@@ -3,157 +3,533 @@
...
+ */
+type FindPartialTuple<T, K extends string[], M extends boolean = false> = K['length'] extends 0
+    ? M extends true
+        ? T
+        : RecursivePartial<T>
+    : {
+          [P in keyof T as P & K[0]]-?: FindPartialTuple<T[P], ReduceTuple<K>, true>
+      } &
+          {
+              [P in keyof T as Exclude<P, K[0]>]+?: FindPartialTuple<T[P], ReduceTuple<K>, false>
```
Comment (kiiadi) [inline src/shared/wizards/wizardForm.ts:111]: whoa - I have no idea what this is doing!

Comment (JadenSimon) [reply src/shared/wizards/wizardForm.ts:111]: Yeah there's a lot of specialized syntax here that's unlike most languages:
* `K['length'] extends 0` is a conditional type to check for an empty tuple
* `M extends true` is there in case we ended on a non-primitive type, in which case we should either pass-through or make it all optional.
* There are two uses of mapped types which have this basic shape: `{ [P in keyof T]: T[P] }` (identity mapping)
  * Mapped types are pretty much what they sound like: they map one type to another. Types in the square brackets are the key, and the right side is the desired type.
* `P in keyof T as P & K[0]` is essentially saying `P` is all keys of `T` that are also the first element of the `K` tuple. The `as` keyword is specifying a mapped key, e.g. `P'(P) = P & K[0]`
* `-?` is a modifier that makes the new type 'non-nullable' (not undefined or null)
* `P in keyof T as Exclude<P, K[0]>` is all keys of `T` that are _not_ the first element of `K`
* `+?` is a modifier that makes the new type 'optional' (so it can be undefined)
* `FindPartialTuple` is the recursive part of the alias. `T[P]` is the type associated with the property key `P` in `T`.
* `ReduceTuple` can be thought of as shifting the tuple, so just removing the first element.

Essentially the alias provides a way to make arbitrary properties of an interface/type required while making everything else optional. In general, it's best to think of Typescript aliases as math functions/definitions considering there's no way to write aliases with an imperative style.
